<?xml version="1.0"?>

<!-- 
BSD 3-Clause License
Copyright (c) 2021, Institut de Robotica i Informatica Industrial (CSIC-UPC)
All rights reserved. 
-->

<launch>
  <arg name="mav_name" default="hexacopter370" />
  <arg name="arm_enable" default="false" />
  <arg name="arm_name" default="flying_arm_3" />
  <arg name="horizon_enable" default="false" />
  <arg name="bag_enable" default="false" />
  <arg name="bag_name" default="" />
  <arg name="bag_player" default="false" />
  <arg name="bag_speed" default="1" unless="$(arg horizon_enable)" />
  <arg name="bag_speed" default="0.2" if="$(arg horizon_enable)" />
  <arg name="payload_enable" default="false" />
  <!-- 0: load -> no load, 1: no load -> load -->
  <arg name="payload_direction" default="0" />
  <arg name="payload_pp_time" default="1.0" />

  <arg name="trajectory_name" default="hover" />
  <arg name="trajectory_namespace" value="trajectory_generation" />
  <arg name="trajectory_config" value="$(arg mav_name)_$(arg trajectory_name).yaml" unless="$(arg arm_enable)" />
  <arg name="trajectory_config" value="$(arg mav_name)_$(arg arm_name)_$(arg trajectory_name).yaml" if="$(arg arm_enable)" />
  <arg name="trajectory_dt" default="20" />
  <arg name="trajectory_solver" default="SolverSbFDDP" />

  <arg name="namespace_controller" value="mpc_controller" />
  <arg name="mpc_config" value="$(arg mav_name)_mpc.yaml" unless="$(arg arm_enable)" />
  <arg name="mpc_config" value="$(arg mav_name)_$(arg arm_name)_mpc.yaml" if="$(arg arm_enable)" />
  <arg name="mpc_type" default="Carrot" />

  <arg name="rviz_file" default="$(find eagle_mpc_viz)/rviz/mpc_controller_display.rviz" unless="$(arg payload_enable)" />
  <arg name="rviz_file" default="$(find eagle_mpc_viz)/rviz/mpc_controller_payload_display.rviz" if="$(arg payload_enable)" />

  <include file="$(find eagle_mpc_viz)/launch/trajectory_generator_viz.launch">
    <arg name="namespace" value="$(arg trajectory_namespace)" />
    <arg name="mav_name" value="$(arg mav_name)" />
    <arg name="arm_name" value="$(arg arm_name)" />
    <arg name="arm_enable" value="$(arg arm_enable)" />
    <arg name="trajectory_name" value="$(arg trajectory_name)" />
    <arg name="trajectory_dt" value="$(arg trajectory_dt)" />
    <arg name="trajectory_solver" value="$(arg trajectory_solver)" />
    <arg name="launch_tools" value="false" />
  </include>

  <group ns="$(arg namespace_controller)">
    <param name="trajectory_path" value="$(find eagle_mpc_yaml)/trajectories/$(arg trajectory_config)" />
    <param name="trajectory_dt" value="$(arg trajectory_dt)" />
    <param name="trajectory_solver" value="$(arg trajectory_solver)" />
    <param name="mpc_path" value="$(find eagle_mpc_yaml)/mpc/$(arg mpc_config)" />
    <param name="mpc_type" value="$(arg mpc_type)" />
    <param name="dynamic_reconfigure_client" value="$(arg trajectory_namespace)/tg_node" />
    <param name="bag_path" value="$(arg bag_name)" if="$(arg bag_enable)" />
    <param name="horizon_enable" value="$(arg horizon_enable)" />
    <param name="payload_enable" value="$(arg payload_enable)" />
    <param name="payload_direction" value="$(arg payload_direction)" if="$(arg payload_enable)" />
    <param name="payload_pp_time" value="$(arg payload_pp_time)" if="$(arg payload_enable)" />

    <node pkg="eagle_mpc_viz" name="mpc_controller_node" type="mpc_controller_viz.py" output="screen" />
    <node type="rviz" name="rviz" pkg="rviz" args="-d $(arg rviz_file)" />
    <node pkg="rqt_reconfigure" type="rqt_reconfigure" name="rqt_reconfigure" unless="$(arg bag_enable)" />
  </group>

  <group if=" $(arg bag_enable)">
    <param name="use_sim_time" value="true" />
    <node pkg="rosbag" type="play" name="player" output="screen" args="$(arg bag_name) -l -r $(arg bag_speed) -s 7" if="$(arg bag_player)" />
  </group>

</launch>